html5lib/_inputstream.py

```diff
@@ -318,45 +360,98 @@ def charsUntil(self, characters, opposite=False):
         characters.
         """
 
-        # Use a cache of regexps to find the required characters
-        try:
-            chars = charsUntilRegEx[(characters, opposite)]
-        except KeyError:
-            if __debug__:
+        if cython.compiled:
+            a_len = cython.declare(Py_ssize_t, index(0x7F) + 1)
+            bitmap = cython.declare(cython.p_ulong)
+            a = cython.declare(array.array)
+            c = cython.declare(cython.Py_UCS4)
+
+            try:
+                a = charsUntilCache[characters]
+            except KeyError:
+                a = array.array('L')
+                array.resize(a, a_len)
+                array.zero(a)
+                bitmap = a.data.as_ulongs
+
                 for c in characters:
-                    assert(ord(c) < 128)
-            regex = "".join(["\\x%02x" % ord(c) for c in characters])
-            if not opposite:
-                regex = "^%s" % regex
-            chars = charsUntilRegEx[(characters, opposite)] = re.compile("[%s]+" % regex)
+                    assert c <= 0x7F
+                    bitmap[index(c)] |= bit(c)
 
-        rv = []
+                charsUntilCache[characters] = a
+            else:
+                bitmap = a.data.as_ulongs
+
+            # we deal with opposite here as it means we can cache purely based
+            # on the pre-existing characters object and not construct a tuple
+            # to also cache on opposite; the below code is much quicker than
+            # constructing a tuple (it'll typically compile to around a dozen
+            # instructions!)
+            if opposite:
+                # this is often too long, but Cython won't let us call sizeof in Pure Python mode
+                new_bitmap = cython.declare(cython.ulong[4])
+                for i in range(a_len):
+                    new_bitmap[i] = ~bitmap[i]
+                bitmap = cython.cast(cython.p_ulong, cython.address(new_bitmap))
+
+            cyrv = ""
 
-        while True:
-            # Find the longest matching prefix
-            m = chars.match(self.chunk, self.chunkOffset)
-            if m is None:
-                # If nothing matched, and it wasn't because we ran out of chunk,
-                # then stop
-                if self.chunkOffset != self.chunkSize:
+            while True:
+                # this really should be a slice of self.chunk, but https://github.com/cython/cython/issues/3536
+                for i in range(self.chunkOffset, self.chunkSize):
+                    with cython.boundscheck(False), cython.wraparound(False):
+                        c = self.chunk[i]
+                    if c > 0x7F and opposite or c <= 0x7F and (bitmap[index(c)] & bit(c)):
+                        cyrv += self.chunk[self.chunkOffset:i]
```

Comment:
ref https://github.com/cython/cython/issues/3887 (IIRC, this was a significant perf issue)